hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvokeOnPartitions.java

```diff
@@ -16,137 +16,212 @@
+            }
+            return;
+        }
+        for (final Map.Entry<Address, List<Integer>> mp : memberPartitions.entrySet()) {
+            final Address address = mp.getKey();
             List<Integer> partitions = mp.getValue();
             PartitionIteratingOperation op = new PartitionIteratingOperation(operationFactory, toIntArray(partitions));
-            Future future = operationService.createInvocationBuilder(serviceName, op, address)
+            operationService.createInvocationBuilder(serviceName, op, address)
                     .setTryCount(TRY_COUNT)
                     .setTryPauseMillis(TRY_PAUSE_MILLIS)
-                    .invoke();
-            futures.put(address, future);
+                    .invoke()
+                    .andThen(new FirstAttemptExecutionCallback(mp.getValue()));
         }
     }
 
-    private void awaitCompletion() {
-        NodeEngineImpl nodeEngine = operationService.nodeEngine;
-        for (Map.Entry<Address, Future> response : futures.entrySet()) {
-            try {
-                Future future = response.getValue();
-                PartitionResponse result = nodeEngine.toObject(future.get());
-                result.addResults(partitionResults);
-            } catch (Throwable t) {
-                if (operationService.logger.isFinestEnabled()) {
-                    operationService.logger.finest(t);
-                } else {
-                    operationService.logger.warning(t.getMessage());
-                }
-                List<Integer> partitions = memberPartitions.get(response.getKey());
-                for (Integer partition : partitions) {
-                    partitionResults.put(partition, t);
+    private void retryPartition(final int partitionId) {
+        Operation operation;
+        PartitionAwareOperationFactory partitionAwareFactory = extractPartitionAware(operationFactory);
+        if (partitionAwareFactory != null) {
+            operation = partitionAwareFactory.createPartitionOperation(partitionId);
+        } else {
+            operation = operationFactory.createOperation();
+        }
+
+        operationService.createInvocationBuilder(serviceName, operation, partitionId)
+                        .invoke()
+                        .andThen(new ExecutionCallback<Object>() {
+                            @Override
+                            public void onResponse(Object response) {
+                                setPartitionResult(partitionId, response);
+                                decrementLatchAndHandle(1);
+                            }
+
+                            @Override
+                            public void onFailure(Throwable t) {
+                                setPartitionResult(partitionId, t);
+                                decrementLatchAndHandle(1);
+                            }
+                        });
+    }
+
+    private void decrementLatchAndHandle(int count) {
```

Comment:
To me it feels like an error-prone situation when we return results for lower number of partitions than requested. It's unclear what such situations indicate: we skipped/forgot some partitions entirely or we processed them without results. We already had such problematic cases recently, EPs were skipping partitions silently and it was really hard to diagnose/debug this.

Not sure what we should do, at least some kind of a logging here and in `FirstAttemptExecutionCallback.onResponse` would be really helpful; a more safe option is to completely disallow such situations by reporting an exceptional result to the future.

Reply:
This behavior was allowed by the old version. I try to make `invokeAsync().get()` equivalent to the old version, that's why I allowed it.

Anyway, I added more checks, as it seems to me that `PartitionIteratingOperation` always returns responses for all partitions.